File: Binary-Tree/BSTIterator/BSTIterator.java
import java.util.LinkedList;

public class BSTIterator{
	public static class TreeNode{
		int val;
		TreeNode left;
		TreeNode right;
		TreeNode(int x){
			this.val=x;
			this.left=null;
			this.right=null;
		}

		public static TreeNode newNode(int val){
			TreeNode node=new TreeNode(val);
			node.left=null;
			node.right=null;
			return node;
		}
		
		public static void printBinaryTree(TreeNode root, int level){
		    if(root==null)
		         return;
		    printBinaryTree(root.right, level+1);
		    if(level!=0){
		        for(int i=0;i<level-1;i++)
		            System.out.print("|\t");
		            System.out.println("|-------"+root.val);
		    }
		    else
		        System.out.println(root.val);
		    printBinaryTree(root.left, level+1);
		}  
	}

	private static LinkedList<TreeNode> stack;
	public BSTIterator(TreeNode root) {
        stack=new LinkedList<TreeNode>();
        TreeNode p=root;
        while(p!=null){
        	stack.push(p);
        	p=p.left;
        }
    }

    /** @return whether we have a next smallest number */
    public static boolean hasNext() {
        return !(stack.size()==0);
    }

    /** @return the next smallest number */
    public static int next() {
        if(!hasNext()) return 0;
        TreeNode node=stack.pop();
        int nodeValue=node.val;
        node=node.right;
        while(node!=null){
        	stack.push(node);
        	node=node.left;
        }
        return nodeValue;
    }

    public static void main(String[] args) {
    	TreeNode root=new TreeNode(20);
    	root.left=new TreeNode(8);
    	root.right=new TreeNode(22);
    	root.left.left=new TreeNode(4);
    	root.left.right=new TreeNode(12);
    	root.left.right.left=new TreeNode(10);
    	root.left.right.right=new TreeNode(14);

    	System.out.println("The BST is: ");
    	TreeNode.printBinaryTree(root,0);

    	BSTIterator it=new BSTIterator(root);
    	while(it.hasNext())
    		System.out.println("Next smallest number is: "+it.next());
    }
}

File: Google/BSTIterator.py
class TreeNode(object):
	def __init__(self,x):
		self.val=x
		self.left=None
		self.right=None

class BSTIterator:
	stack=[]
	def __init__(self, root):
		"""
		:type root: TreeNode
		"""
		p=root
		while p is not None:
			self.stack.append(p)
			p=p.left

	def hasNext(self):
		"""
		:rtype: bool
		"""
		return True if len(self.stack)!=0 else False

	def next(self):
		"""
		:rtype: int
		"""
		if len(self.stack)==0: return 0
		node=self.stack.pop()
		res=node.val
		node=node.right
		while node is not None:
			self.stack.append(node)
			node=node.left
		return res

root=None
def insert(x):
	"""
		type: x -- int
		rType: void
	"""
	global root
	root=create_node(root,x)

def create_node(root,x):
	if root is None:
		return TreeNode(x)
	if x<=root.val:
		root.left=create_node(root.left,x)
	else:
		root.right=create_node(root.right,x)
	return root

def print_tree():
	global root
	print_helper(root, "")

def print_helper(root, indent):
	if root is not None:
		print_helper(root.right, indent+"   ")
		print indent+str(root.val)
		print_helper(root.left, indent+"   ")


insert(7)
insert(3)
insert(11)
insert(13)
insert(1)
insert(5)
insert(9)
print_tree()

it,v=BSTIterator(root),[]
while(it.hasNext()):
	v.append(it.next())
print v

